chromium: Add PACKAGECONFIG settings to switch between GTK 3/4

zboszor opened the pull request:
Kirkstone provides gtk4 4.4.0, Chromium 97 has a configuration
setting to build with GTK version 3 or 4. Expose this choice in
PACKAGECONFIG.

Signed-off-by: Zoltán Böszörményi <zboszor@gmail.com>

rakuco commented:
I think @msisov knows more about this upstream, but IIRC Chromium could choose at runtime whether to use GTK3 or GTK4. Is the current code causing problems in kirkstone?

zboszor commented:
> I think @msisov knows more about this upstream, but IIRC Chromium could choose at runtime whether to use GTK3 or GTK4. Is the current code causing problems in kirkstone?

I don't know about runtime selection but if I chose to build with GTK4, then build errors occurred. I was just observing them and added the patch in the second commit which fixed those compiler errors.

msisov commented:
Please see https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk_compat.cc;drc=ecc7a574794f6b2a94b4047b6648ea118b0217f9;bpv=1;bpt=1;l=122?q=gtk-version&ss=chromium%2Fchromium%2Fsrc .

Chromium dlopens the requested library then - https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk_compat.cc;l=79;drc=ecc7a574794f6b2a94b4047b6648ea118b0217f9;bpv=1;bpt=1?q=gtk-version&ss=chromium%2Fchromium%2Fsrc

zboszor commented:
> Please see https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk_compat.cc;drc=ecc7a574794f6b2a94b4047b6648ea118b0217f9;bpv=1;bpt=1;l=122?q=gtk-version&ss=chromium%2Fchromium%2Fsrc .
> 
> Chromium dlopens the requested library then - https://source.chromium.org/chromium/chromium/src/+/main:ui/gtk/gtk_compat.cc;l=79;drc=ecc7a574794f6b2a94b4047b6648ea118b0217f9;bpv=1;bpt=1?q=gtk-version&ss=chromium%2Fchromium%2Fsrc

Nice, but it doesn't contradict with the fact that if I set gtk_version=4 for the build, I get build errors:
```
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:1:19: error: conflicting types for 'gsk_render_node_get_node_type'
| GskRenderNodeType gsk_render_node_get_node_type(GskRenderNode* node);
|                   ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:121:25: note: previous declaration is here
| GskRenderNodeType       gsk_render_node_get_node_type           (const GskRenderNode *node);
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:3:16: error: conflicting types for 'gsk_transform_node_get_child'
| GskRenderNode* gsk_transform_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:389:25: note: previous declaration is here
| GskRenderNode *         gsk_transform_node_get_child            (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:4:16: error: conflicting types for 'gsk_opacity_node_get_child'
| GskRenderNode* gsk_opacity_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:399:25: note: previous declaration is here
| GskRenderNode *         gsk_opacity_node_get_child              (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:5:16: error: conflicting types for 'gsk_color_matrix_node_get_child'
| GskRenderNode* gsk_color_matrix_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:410:25: note: previous declaration is here
| GskRenderNode *         gsk_color_matrix_node_get_child         (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:6:16: error: conflicting types for 'gsk_repeat_node_get_child'
| GskRenderNode* gsk_repeat_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:424:25: note: previous declaration is here
| GskRenderNode *         gsk_repeat_node_get_child               (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:7:16: error: conflicting types for 'gsk_clip_node_get_child'
| GskRenderNode* gsk_clip_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:434:25: note: previous declaration is here
| GskRenderNode *         gsk_clip_node_get_child                 (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:8:16: error: conflicting types for 'gsk_rounded_clip_node_get_child'
| GskRenderNode* gsk_rounded_clip_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:444:25: note: previous declaration is here
| GskRenderNode *         gsk_rounded_clip_node_get_child         (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:9:16: error: conflicting types for 'gsk_shadow_node_get_child'
| GskRenderNode* gsk_shadow_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:455:25: note: previous declaration is here
| GskRenderNode *         gsk_shadow_node_get_child               (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:10:16: error: conflicting types for 'gsk_blur_node_get_child'
| GskRenderNode* gsk_blur_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:515:25: note: previous declaration is here
| GskRenderNode *         gsk_blur_node_get_child                 (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:11:16: error: conflicting types for 'gsk_debug_node_get_child'
| GskRenderNode* gsk_debug_node_get_child(GskRenderNode* node);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:200:25: note: previous declaration is here
| GskRenderNode *         gsk_debug_node_get_child                (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:12:16: error: conflicting types for 'gsk_container_node_get_child'
| GskRenderNode* gsk_container_node_get_child(GskRenderNode* node, guint idx);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:380:25: note: previous declaration is here
| GskRenderNode *         gsk_container_node_get_child            (const GskRenderNode      *node,
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:13:16: error: conflicting types for 'gsk_gl_shader_node_get_child'
| GskRenderNode* gsk_gl_shader_node_get_child(GskRenderNode* node, guint idx);
|                ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:530:25: note: previous declaration is here
| GskRenderNode *         gsk_gl_shader_node_get_child            (const GskRenderNode      *node,
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:14:7: error: conflicting types for 'gsk_container_node_get_n_children'
| guint gsk_container_node_get_n_children(GskRenderNode* node);
|       ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:378:25: note: previous declaration is here
| guint                   gsk_container_node_get_n_children       (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:15:7: error: conflicting types for 'gsk_gl_shader_node_get_n_children'
| guint gsk_gl_shader_node_get_n_children(GskRenderNode* node);
|       ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:528:25: note: previous declaration is here
| guint                   gsk_gl_shader_node_get_n_children       (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| In file included from gen/ui/gtk/gtk_stubs.cc:12:
| In file included from ../../ui/gtk/gtk_compat.h:28:
| ../../ui/gtk/gsk.sigs:16:13: error: conflicting types for 'gsk_texture_node_get_texture'
| GdkTexture* gsk_texture_node_get_texture(GskRenderNode* node);
|             ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:218:25: note: previous declaration is here
| GdkTexture *            gsk_texture_node_get_texture            (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| gen/ui/gtk/gtk_stubs.cc:642:26: error: conflicting types for 'gsk_render_node_get_node_type'
| extern GskRenderNodeType gsk_render_node_get_node_type(GskRenderNode* node) __attribute__((weak));
|                          ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:121:25: note: previous declaration is here
| GskRenderNodeType       gsk_render_node_get_node_type           (const GskRenderNode *node);
|                         ^
| gen/ui/gtk/gtk_stubs.cc:644:20: error: conflicting types for 'gsk_render_node_get_node_type'
| GskRenderNodeType  gsk_render_node_get_node_type(GskRenderNode* node) {
|                    ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:121:25: note: previous declaration is here
| GskRenderNodeType       gsk_render_node_get_node_type           (const GskRenderNode *node);
|                         ^
| gen/ui/gtk/gtk_stubs.cc:652:23: error: conflicting types for 'gsk_transform_node_get_child'
| extern GskRenderNode* gsk_transform_node_get_child(GskRenderNode* node) __attribute__((weak));
|                       ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:389:25: note: previous declaration is here
| GskRenderNode *         gsk_transform_node_get_child            (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| gen/ui/gtk/gtk_stubs.cc:654:17: error: conflicting types for 'gsk_transform_node_get_child'
| GskRenderNode*  gsk_transform_node_get_child(GskRenderNode* node) {
|                 ^
| /data/genericx86-64-yocto-3.4/tmp-sicom-glibc/work/core2-64-oe-linux/chromium-x11/97.0.4692.71-r0/recipe-sysroot/usr/include/gtk-4.0/gsk/gskrendernode.h:389:25: note: previous declaration is here
| GskRenderNode *         gsk_transform_node_get_child            (const GskRenderNode      *node) G_GNUC_PURE;
|                         ^
| fatal error: too many errors emitted, stopping now [-ferror-limit=]
| 20 errors generated.
```
Then:
```
| ../../ui/gtk/gtk_util.cc:710:28: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_TRANSFORM_NODE, gsk_transform_node_get_child},
|                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:711:26: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_OPACITY_NODE, gsk_opacity_node_get_child},
|                          ^~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:712:31: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_COLOR_MATRIX_NODE, gsk_color_matrix_node_get_child},
|                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:713:25: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_REPEAT_NODE, gsk_repeat_node_get_child},
|                         ^~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:714:23: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_CLIP_NODE, gsk_clip_node_get_child},
|                       ^~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:715:31: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_ROUNDED_CLIP_NODE, gsk_rounded_clip_node_get_child},
|                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:716:25: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_SHADOW_NODE, gsk_shadow_node_get_child},
|                         ^~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:717:23: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_BLUR_NODE, gsk_blur_node_get_child},
|                       ^~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:718:24: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *)' (aka '_GskRenderNode *(*)(_GskRenderNode *)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *)' (aka '_GskRenderNode *(const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_DEBUG_NODE, gsk_debug_node_get_child},
|                        ^~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:725:28: error: cannot initialize a member subobject of type 'guint (*)(GskRenderNode *)' (aka 'unsigned int (*)(_GskRenderNode *)') with an lvalue of type 'guint (const GskRenderNode *)' (aka 'unsigned int (const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_CONTAINER_NODE, gsk_container_node_get_n_children,
|                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:726:8: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *, guint)' (aka '_GskRenderNode *(*)(_GskRenderNode *, unsigned int)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *, guint)' (aka '_GskRenderNode *(const _GskRenderNode *, unsigned int)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|        gsk_container_node_get_child},
|        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:727:28: error: cannot initialize a member subobject of type 'guint (*)(GskRenderNode *)' (aka 'unsigned int (*)(_GskRenderNode *)') with an lvalue of type 'guint (const GskRenderNode *)' (aka 'unsigned int (const _GskRenderNode *)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|       {GSK_GL_SHADER_NODE, gsk_gl_shader_node_get_n_children,
|                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
| ../../ui/gtk/gtk_util.cc:728:8: error: cannot initialize a member subobject of type 'GskRenderNode *(*)(GskRenderNode *, guint)' (aka '_GskRenderNode *(*)(_GskRenderNode *, unsigned int)') with an lvalue of type 'GskRenderNode *(const GskRenderNode *, guint)' (aka '_GskRenderNode *(const _GskRenderNode *, unsigned int)'): type mismatch at 1st parameter ('GskRenderNode *' (aka '_GskRenderNode *') vs 'const GskRenderNode *' (aka 'const _GskRenderNode *'))
|        gsk_gl_shader_node_get_child},
|        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
| 13 errors generated.
```

rakuco commented:
> > I think @msisov knows more about this upstream, but IIRC Chromium could choose at runtime whether to use GTK3 or GTK4. Is the current code causing problems in kirkstone?
> 
> I don't know about runtime selection but if I chose to build with GTK4, then build errors occurred. I was just observing them and added the patch in the second commit which fixed those compiler errors.

Right, the build errors do need to be fixed either here or upstream (preferably), I think the question is whether it makes sense to add the PACKAGECONFIG knob and allow building with GTK4 instead of GTK3 given that not doing that still allows Chromium to run with GTK4 if it's present and GTK3 is not. Is the current setup failing for you somehow?

zboszor commented:
> > > I think @msisov knows more about this upstream, but IIRC Chromium could choose at runtime whether to use GTK3 or GTK4. Is the current code causing problems in kirkstone?
> > 
> > 
> > I don't know about runtime selection but if I chose to build with GTK4, then build errors occurred. I was just observing them and added the patch in the second commit which fixed those compiler errors.
> 
> Right, the build errors do need to be fixed either here or upstream (preferably), I think the question is whether it makes sense to add the PACKAGECONFIG knob and allow building with GTK4 instead of GTK3 given that not doing that still allows Chromium to run with GTK4 if it's present and GTK3 is not. Is the current setup failing for you somehow?

For one, it fails to pull in `gtk4` as a runtime dependency if I want to run Chromium that way because of `dlopen()`-ing the library.  The packageconfig knob also adds the runtime dependency.

zboszor commented:
> For one, it fails to pull in `gtk4` as a runtime dependency if I want to run Chromium that way because of `dlopen()`-ing the library. The packageconfig knob also adds the runtime dependency.

Also, it still uses GTK3 even if it was built with GTK4 and GTK4 is present on the system. It only uses GTK4 if GTK3 is not present. So, yes, I would say the current setup is failing to satisfy the expectations.

zboszor commented:
I think the problem is unfolding now: `inherit gtk-icon-cache` inherently adds `gtk+3` to `RDEPENDS`.
Kirkstone may need a new `gtk4-icon-cache.bbclass` to be used by recipes wanting to use `gtk4`.
I have previously discovered such an inherit line does work:
```
inherit ${@bb.utils.contains('PACKAGECONFIG', 'gtk4', 'gtk4-icon-cache', 'gtk-icon-cache', d)}
```
Also the wrapper script  needs the extra command line option `--gtk-version=4`

zboszor commented:
I have removed the extra patch to build with GTK4 >= 4.2.0 and reverted to build with GTK3.
The PR now needs this change so RDEPENDS will use gtk4 instead of gtk+3:
https://lists.openembedded.org/g/openembedded-core/message/162302

zboszor commented:
@rakuco @msisov My change for `gtk-icon-cache.bbclass` was merged into openembedded-core master.
Can you please consider merging this PR now? Thanks in advance.

msisov reviewed (approved):
i think it's fine. @rakuco wdyt?

rakuco commented on meta-chromium/recipes-browser/chromium/chromium-gn.inc:
```diff
@@ -116,6 +118,7 @@ PACKAGECONFIG[use-egl] = ",,virtual/egl virtual/libgles2"
 # and command-line switches (extra dependencies should not
 # be necessary but are OK to add).
 PACKAGECONFIG[component-build] = ""
+PACKAGECONFIG[gtk4] = ""
```
- This needs to be documented in `meta-chromium/README.md`.
- Can you move it a couple of lines below so the list remains sorted?

zboszor replied:
For your bulleted questions above, the answer to all of them are yes. I will make the requested changes soon. Thanks for your review.

zboszor replied:
BTW, it's `gtk-icon-cache.bbappend` that adds the build and runtime dependency on `gtk+3{-native}`, and on `gtk4` with the PACKAGECONFIG knob set.

zboszor replied:
For the question in bullet number 3: it works with older meta-openembedded branches, assuming that either

- `PACKAGECONFIG=gtk4` is not used, or
- someone has a `gtk4` recipe plus the `gtk-icon-cache.bbclass` change backported to the older branch

rakuco replied:
> BTW, it's `gtk-icon-cache.bbappend` that adds the build and runtime dependency on `gtk+3{-native}`, and on `gtk4{-native}` with the PACKAGECONFIG knob set.

Does it mean we can remove the explicit `DEPENDS` on `gtk+3` from `chromium-gn.inc`?

zboszor replied:
The explicit `DEPENDS` on `gtk+3` can be omitted from `chromium-gn.inc` because `gtk-icon-cache.bbclass` adds it.

rakuco reviewed (commented):
Thanks for working on this, I'm glad to see that the end result is smaller than the first version of this patch :-)

Let me see if I've understood it correctly:
- It's OK to continue depending on and building against `gtk+3` even if people want to use GTK+4 at runtime
- It is also OK not to have any dependency on gtk4 even if people set the `gtk4` `PACKAGECONFIG` knob
- This also works with older meta-openembedded branches

rakuco reviewed (approved):
lgtm, thanks for driving this work and answering all my questions!
